src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/RazorProjectFileSystem.cs

```diff
@@ -88,45 +89,72 @@ public virtual IEnumerable<RazorProjectItem> FindHierarchicalItems(string basePa
 
         if (path.Length == 1)
         {
-            yield break;
+            return [];
         }
 
         if (!path.StartsWith(basePath, StringComparison.OrdinalIgnoreCase))
         {
-            yield break;
+            return [];
         }
 
-        StringBuilder builder;
         var fileNameIndex = path.LastIndexOf('/');
-        var length = path.Length;
 
         if (fileNameIndex == -1)
         {
             throw new InvalidOperationException($"Cannot find file name in path '{path}'");
         }
 
-        if (string.Compare(path, fileNameIndex + 1, fileName, 0, fileName.Length, StringComparison.Ordinal) == 0)
-        {
-            // If the specified path is for the file hierarchy being constructed, then the first file that applies
-            // to it is in a parent directory.
-            builder = new StringBuilder(path, 0, fileNameIndex, fileNameIndex + fileName.Length);
-            length = fileNameIndex;
-        }
-        else
+        var length = fileNameIndex + 1;
+        var pathMemory = path.AsMemory();
+
+        if (pathMemory.Span[(fileNameIndex + 1)..].Equals(fileName.AsSpan(), StringComparison.Ordinal))
         {
-            builder = new StringBuilder(path);
+            pathMemory = pathMemory[..fileNameIndex];
         }
 
-        var maxDepth = 255;
-        var index = length;
-        while (maxDepth-- > 0 && index > basePath.Length && (index = path.LastIndexOf('/', index - 1)) != -1)
+        using var result = new PooledArrayBuilder<RazorProjectItem>();
+
+        const int MaxDepth = 255;
```

Comment:
Any consideration to just "let's break it"?

Reply:
It was easy enough to keep the existing behavior, but I'm happy to remove it if you prefer. The idea that we need a check to ensure that we don't ever consider imports outside of (2^8) - 1 subdirectories is definitely silly!

Reply:
I don't know the reason for [this fence](https://fs.blog/chestertons-fence/), but if I had to guesss I would assume it is probably to prevent recursive directories via sym/hardlinks, so we don't get stuck in an infinite loop looking for imports.

Reply:
I agree with that sentiment, but this doesn't actually recurse directories. It just builds up path strings by walking through the slashes.

Reply:
Hah, well then, my vote is change it.

Reply:
Okey-dokey. That gives me enough plausible deniability. :smile: